Week-3/6_CrowdFunding/Crowdfunding.sol

```diff
@@ -0,0 +1,153 @@
+
+    allProjects.push(newProject);
+  }
+
+  /**
+   * The function to contribute funds to a particular project
+   * @param _projectId The id of the project you wanr to make donations to
+   */
+  function contributeFunds(uint256 _projectId)
+    external
+    payable
+    isValidProject(_projectId)
+  {
+    require(msg.value > 0, "Please enter a valid amount");
+
+    Project memory requiredProject = allProjects[_projectId];
+
+    require(
+      requiredProject.amountNeeded - requiredProject.fundingReceived >=
+        msg.value,
+      "Not needed this much amount"
+    );
+
+    allProjects[_projectId].fundingReceived += msg.value;
+    contributors[_projectId][msg.sender] += msg.value;
+  }
+
+  /**
+   * The function to claim funds by the creator
+   * @param _projectId The id of the project from which the claim is to be made
+   * @param _amountToBeClaimed The amount that is to be claimed
+   */
+  function claimFunds(uint256 _projectId, uint256 _amountToBeClaimed)
+    external
+    payable
+    isValidProject(_projectId)
+    isProjectOwner(_projectId)
+  {
+    Project memory requiredProject = allProjects[_projectId];
+    require(!requiredProject.completeAmountClaimed, "Funds already claimed");
+
+    require(
+      _amountToBeClaimed <= requiredProject.fundingReceived,
+      "Insufficient Funds"
+    );
+
+    (bool success, ) = requiredProject.raisedFor.call{
+      value: _amountToBeClaimed
+    }("");
+
+    require(success, "Something went wrong. Please try after some time.");
+
+    allProjects[_projectId].amountClaimed += _amountToBeClaimed;
+
+    if (
+      requiredProject.amountClaimed + _amountToBeClaimed ==
+      requiredProject.amountNeeded
+    ) {
+      allProjects[_projectId].completeAmountClaimed = true;
+    }
+  }
+
+  /**
+   *
+   * @param _projectId The id of the project from  which the funds are to be withdrawn
+   * @param _amountToBeWithdrawn The amount that is to be withdrawn
+   */
+  function withdrawFunds(uint256 _projectId, uint256 _amountToBeWithdrawn)
+    external
+    payable
+    isValidProject(_projectId)
+  {
+    Project memory requiredProject = allProjects[_projectId];
+
+    require(!requiredProject.completeAmountClaimed, "Funds have been claimed");
+
+    // amount can be withdrawn on;y if the remaining amount in the project >= amount to be withdrawn
+    require(
+      _amountToBeWithdrawn > 0 &&
+        requiredProject.fundingReceived - requiredProject.amountClaimed > 0 &&
+        _amountToBeWithdrawn <=
+        requiredProject.fundingReceived - requiredProject.amountClaimed &&
+        _amountToBeWithdrawn <= contributors[_projectId][msg.sender],
+      "Insufficient Balance"
+    );
+
+    (bool success, ) = msg.sender.call{value: _amountToBeWithdrawn}("");
+
+    require(success, "Something went wrong. Please try after some time.");
+
+    contributors[_projectId][msg.sender] -= _amountToBeWithdrawn;
+    allProjects[_projectId].fundingReceived -= _amountToBeWithdrawn;
```

Comment:
This is prone to reentrancy attack -

If msg.sender is a contract, it could in theory call withdrawFunds again when it receives the Ether, and since the balances aren't updated yet, it could potentially withdraw more funds than it should.

Use the Checks-Effects-Interactions pattern in your code to avoid this. Please read more here: https://consensys.github.io/smart-contract-best-practices/attacks/reentrancy/

Reply:
change done.
Also updating balance in case the transfer fails